File: csharp/IfaUdi.Parser/Envelope.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using IfaUdi.Parser.Models;

namespace IfaUdi.Parser
{
    /// <summary>
    /// Detects and normalizes the three accepted IFA UDI barcode envelope forms (raw ISO/IEC 15434
    /// Format 06, DIN 16598 keyboard-compatible, and the printable Interpretation Line) into an
    /// ordered list of (DataIdentifier, Value) pairs. See documentation/IFA_UDI_Parser_Analysis.md
    /// section 4.
    /// </summary>
    internal static class Envelope
    {
        private const char RecordSeparator = (char)0x1E;
        private const char GroupSeparator = (char)0x1D;
        private const char EndOfTransmission = (char)0x04;

        // Known Data Identifiers, longest first so prefix matching is unambiguous.
        private static readonly string[] KnownDataIdentifiers =
            new[] { "27Q", "16D", "33L", "9N", "1T", "8P", "D", "S", "Q" };

        private static readonly Regex InterpretationLineFieldPattern = new Regex(@"\(([0-9A-Za-z]+)\)([^()]*)", RegexOptions.Compiled);

        public static List<(string Di, string Value)> Normalize(string barcode)
        {
            if (string.IsNullOrEmpty(barcode))
            {
                throw new IfaUdiFormatException("Barcode input must not be empty.");
            }

            if (barcode.StartsWith("[)>", StringComparison.Ordinal))
            {
                return ParseRawIso15434(barcode);
            }

            if (barcode[0] == '.')
            {
                return ParseDin16598(barcode);
            }

            if (barcode[0] == '(' && InterpretationLineFieldPattern.IsMatch(barcode))
            {
                return ParseInterpretationLine(barcode);
            }

            throw new IfaUdiFormatException("Barcode does not match any recognized IFA UDI envelope form (raw ISO/IEC 15434, DIN 16598, or Interpretation Line).");
        }

        private static List<(string Di, string Value)> ParseRawIso15434(string barcode)
        {
            string rest = barcode.Substring(3); // strip "[)>"

            if (rest.Length > 0 && rest[0] == RecordSeparator)
            {
                rest = rest.Substring(1);
            }

            if (!rest.StartsWith("06", StringComparison.Ordinal))
            {
                throw new IfaUdiFormatException("Raw ISO/IEC 15434 envelope must use Format 06.");
            }

            rest = rest.Substring(2);

            if (rest.Length > 0 && rest[0] == GroupSeparator)
            {
                rest = rest.Substring(1);
            }

            rest = rest.TrimEnd(RecordSeparator, EndOfTransmission);

            string[] rawFields = rest.Split(GroupSeparator);
            return SplitDataIdentifiers(rawFields);
        }

        private static List<(string Di, string Value)> ParseDin16598(string barcode)
        {
            string rest = barcode.Substring(1); // strip leading '.'
            string[] rawFields = rest.Split('^');
            return SplitDataIdentifiers(rawFields);
        }

        private static List<(string Di, string Value)> ParseInterpretationLine(string barcode)
        {
            var result = new List<(string, string)>();
            foreach (Match match in InterpretationLineFieldPattern.Matches(barcode))
            {
                result.Add((match.Groups[1].Value, match.Groups[2].Value));
            }

            return result;
        }

        public static string Serialize(List<(string Di, string Value)> fields, EnvelopeForm form)
        {
            switch (form)
            {
                case EnvelopeForm.InterpretationLine:
                    return string.Concat(fields.Select(f => $"({f.Di}){f.Value}"));
                case EnvelopeForm.RawIso15434:
                    return $"[)>{RecordSeparator}06{GroupSeparator}" +
                        string.Join(GroupSeparator.ToString(), fields.Select(f => f.Di + f.Value)) +
                        $"{RecordSeparator}{EndOfTransmission}";
                case EnvelopeForm.Din16598:
                    return "." + string.Join("^", fields.Select(f => f.Di + f.Value));
                default:
                    throw new ArgumentOutOfRangeException(nameof(form));
            }
        }

        private static List<(string Di, string Value)> SplitDataIdentifiers(string[] rawFields)
        {
            var result = new List<(string, string)>();
            foreach (string field in rawFields)
            {
                string di = KnownDataIdentifiers.FirstOrDefault(field.StartsWith);
                if (di is null)
                {
                    throw new IfaUdiFormatException($"Field '{field}' does not start with a recognized Data Identifier.");
                }

                result.Add((di, field.Substring(di.Length)));
            }

            return result;
        }
    }
}

File: typescript/ifa-udi-parser/src/envelope.ts
import { IfaUdiFormatError, type EnvelopeForm } from './types.js';

/**
 * Detects and normalizes the three accepted IFA UDI barcode envelope forms (raw ISO/IEC 15434
 * Format 06, DIN 16598 keyboard-compatible, and the printable Interpretation Line) into an
 * ordered list of (DataIdentifier, Value) pairs. See documentation/IFA_UDI_Parser_Analysis.md
 * section 4.
 */

const RECORD_SEPARATOR = String.fromCharCode(0x1e);
const GROUP_SEPARATOR = String.fromCharCode(0x1d);
const END_OF_TRANSMISSION = String.fromCharCode(0x04);

// Known Data Identifiers, longest first so prefix matching is unambiguous.
const KNOWN_DATA_IDENTIFIERS = ['27Q', '16D', '33L', '9N', '1T', '8P', 'D', 'S', 'Q'];

const INTERPRETATION_LINE_FIELD_PATTERN = /\(([0-9A-Za-z]+)\)([^()]*)/g;

export interface EnvelopeField {
  di: string;
  value: string;
}

export function normalize(barcode: string): EnvelopeField[] {
  if (!barcode) {
    throw new IfaUdiFormatError('Barcode input must not be empty.');
  }

  if (barcode.startsWith('[)>')) {
    return parseRawIso15434(barcode);
  }

  if (barcode.startsWith('.')) {
    return parseDin16598(barcode);
  }

  if (barcode.startsWith('(') && new RegExp(INTERPRETATION_LINE_FIELD_PATTERN).test(barcode)) {
    return parseInterpretationLine(barcode);
  }

  throw new IfaUdiFormatError(
    'Barcode does not match any recognized IFA UDI envelope form (raw ISO/IEC 15434, DIN 16598, or Interpretation Line).',
  );
}

function parseRawIso15434(barcode: string): EnvelopeField[] {
  let rest = barcode.slice(3); // strip "[)>"

  if (rest.startsWith(RECORD_SEPARATOR)) {
    rest = rest.slice(1);
  }

  if (!rest.startsWith('06')) {
    throw new IfaUdiFormatError('Raw ISO/IEC 15434 envelope must use Format 06.');
  }

  rest = rest.slice(2);

  if (rest.startsWith(GROUP_SEPARATOR)) {
    rest = rest.slice(1);
  }

  while (rest.endsWith(RECORD_SEPARATOR) || rest.endsWith(END_OF_TRANSMISSION)) {
    rest = rest.slice(0, -1);
  }

  const rawFields = rest.split(GROUP_SEPARATOR);
  return splitDataIdentifiers(rawFields);
}

function parseDin16598(barcode: string): EnvelopeField[] {
  const rest = barcode.slice(1); // strip leading '.'
  const rawFields = rest.split('^');
  return splitDataIdentifiers(rawFields);
}

function parseInterpretationLine(barcode: string): EnvelopeField[] {
  const result: EnvelopeField[] = [];
  const pattern = new RegExp(INTERPRETATION_LINE_FIELD_PATTERN);
  let match: RegExpExecArray | null;
  while ((match = pattern.exec(barcode)) !== null) {
    result.push({ di: match[1], value: match[2] });
  }

  return result;
}

export function serialize(fields: EnvelopeField[], form: EnvelopeForm): string {
  switch (form) {
    case 'interpretationLine':
      return fields.map((f) => `(${f.di})${f.value}`).join('');
    case 'rawIso15434':
      return (
        `[)>${RECORD_SEPARATOR}06${GROUP_SEPARATOR}` +
        fields.map((f) => `${f.di}${f.value}`).join(GROUP_SEPARATOR) +
        `${RECORD_SEPARATOR}${END_OF_TRANSMISSION}`
      );
    case 'din16598':
      return '.' + fields.map((f) => `${f.di}${f.value}`).join('^');
  }
}

function splitDataIdentifiers(rawFields: string[]): EnvelopeField[] {
  return rawFields.map((field) => {
    const di = KNOWN_DATA_IDENTIFIERS.find((candidate) => field.startsWith(candidate));
    if (di === undefined) {
      throw new IfaUdiFormatError(`Field '${field}' does not start with a recognized Data Identifier.`);
    }

    return { di, value: field.slice(di.length) };
  });
}
